src/Timelock.sol

```diff
@@ -0,0 +1,739 @@
+        return timestamps[id];
+    }
+
+    /// @dev Returns the identifier of an operation containing a single transaction.
+    function hashOperation(
+        address target,
+        uint256 value,
+        bytes calldata data,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public pure virtual returns (bytes32) {
+        return keccak256(abi.encode(target, value, data, predecessor, salt));
+    }
+
+    /// @dev Returns the identifier of an operation containing a batch of transactions.
+    function hashOperationBatch(
+        address[] calldata targets,
+        uint256[] calldata values,
+        bytes[] calldata payloads,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public pure virtual returns (bytes32) {
+        return
+            keccak256(abi.encode(targets, values, payloads, predecessor, salt));
+    }
+
+    /// ---------------------------------------------------------------
+    /// ---------------------------------------------------------------
+    /// -------------------- Timelock Functions -----------------------
+    /// ---------------------------------------------------------------
+    /// ---------------------------------------------------------------
+
+    /// @dev Schedule an operation containing a single transaction.
+    /// Emits {CallSalt} if salt is nonzero, and {CallScheduled}.
+    /// Requirements:
+    ///   the caller must have the 'proposer' role.
+    function schedule(
+        address target,
+        uint256 value,
+        bytes calldata data,
+        bytes32 predecessor,
+        bytes32 salt,
+        uint256 delay
+    ) public virtual onlySafe whenNotPaused {
+        bytes32 id = hashOperation(target, value, data, predecessor, salt);
+
+        require(_liveProposals.add(id), "failed to add proposal, duplicate id");
+
+        _schedule(id, delay);
+
+        emit CallScheduled(
+            id,
+            0,
+            target,
+            value,
+            data,
+            predecessor,
+            salt,
+            delay
+        );
+    }
+
+    /// @dev Schedule an operation containing a batch of transactions.
+    /// Emits {CallSalt} if salt is nonzero, and one {CallScheduled} event per transaction in the batch.
+    /// Requirements:
+    /// - the caller must be the safe
+    /// @param targets the addresses of the contracts to call
+    /// @param values the values to send in the calls
+    /// @param payloads the calldata to send in the calls
+    /// @param predecessor the id of the predecessor operation
+    /// @param salt the salt to be used in the operation
+    /// @param delay the delay before the operation becomes valid
+    function scheduleBatch(
+        address[] calldata targets,
+        uint256[] calldata values,
+        bytes[] calldata payloads,
+        bytes32 predecessor,
+        bytes32 salt,
+        uint256 delay
+    ) public virtual onlySafe whenNotPaused {
```

Comment:
why no call to checkCalldata here?

Reply:
I guess because any calldata can be executed as long is using the timelock delay?